test_pull_requests

```diff
@@ -152,6 +178,97 @@ def submitted_tests_for_branch(branch)
   $submitted_tests[branch] ? $submitted_tests[branch] : $submitted_tests['*']
 end
 
+# Pulls out as much metadata from the bot comment as it can, including
+# state (AKA content), Jenkins build URL, base repo commit ID, and the
+# extended tests metadata
+def extract_bot_comment_fields(bot_comment_body, settings=nil)
+  test_prefix = nil
+  content = nil
+  build_url = nil
+  base_commit = nil
+  extended_tests = nil
+  state = nil
+  build_url_pattern = nil
+  if settings.nil?
+    # Use this test to set the build url matching pattern
+    if ENV['BUILD_URL']
+      # Don't set build_url directly from the env var, since we want
+      # to make sure they match
+      build_url_pattern = / \((#{ENV['BUILD_URL']})\)/
+    else
+      build_url_pattern = / \((https?:[^)]+)\)/
+    end
+    # Do our best to find the prefix since we don't know it beforehand
+    if bot_comment_body =~ /^([^:(]+): /
+      test_prefix = $1.strip
+    elsif bot_comment_body =~ /^([^ (]+) /
+      test_prefix = $1.strip
+    end
+  else
+    build_url_pattern = / \((#{settings['jenkins_url']}[^)]+)\)/
+    test_prefix = settings['test_prefix']
+  end
+  # Do our best to find the content (aka state) even if there aren't metadata in parens after it
+  if test_prefix && bot_comment_body =~ /^#{test_prefix} ([^(]+) ?\(?/
+    content = $1.strip
+    # Match up to colon or space since "Waiting" content varies
+    state = CONTENT_TO_STATE[content[/^(#{CONTENT_EVALUATING}|[^: ]+)/, 1]]
+  end
+  build_url = bot_comment_body[build_url_pattern, 1]
+  base_commit = bot_comment_body[/ (\(Base Commit: [a-f0-9]{40}\))/, 1]
+  extended_tests = bot_comment_body[/ (\(Extended Tests: .+)/, 1]
+  if content.nil?
+    if settings.nil?
+      raise "Could not parse bot comment: '#{bot_comment_body}'"
+    end
+    raise "'#{settings['name']}' pull request not found"
+  end
+
+  return {
+    :prefix => test_prefix,
+    :content => content,
+    :build_url => build_url,
+    :base_commit => base_commit,
+    :extended_tests => extended_tests,
+    :state => state
+  }
+end
+
+# Constructs a "standard" format bot comment from the provided test
+# settings prefix, comment, and metadata. Uses BOT_COMMENT_FIELDS to
+# map field keys to field prefixes in the comment string
+def compose_bot_comment(prefix, comment_data={})
+  # Derive comment data content from state, if needed
+  if comment_data[:state] && (comment_data[:content].nil? ||
+                              !comment_data[:content].start_with?(STATE_TO_CONTENT[comment_data[:state]]))
+    comment_data[:content] = STATE_TO_CONTENT[comment_data[:state]]
+  end
+  updated_comment = "#{prefix} #{comment_data[:content]}"
+  BOT_COMMENT_FIELDS.each do |field_key, field_prefix|
+    if comment_data[field_key] && (!comment_data[field_key].empty?)
+      field_data = comment_data[field_key]
+      updated_comment += if field_data.start_with? '('
```

Comment:
This kind of construct is one of the few things I like about ruby. It's clear that the assignment is the result of the logic on the right. If the logic were any more verbose I would be happy to break it out.